### PR TITLE
Prometheus: Auto legend handling 

### PR DESCRIPTION
**Pull request by torkelo**
Progress on #44253

* Introduces a special legendFormat state of `__auto`  , when in this state and if there are labels the backend will not return any name in DisplayNameFromDS. This means the frontend automatic naming will kick in which will try to not show duplicate information. For example if there is only one label then it will skip showing the label name. Example for series names like `{status_code=403}` if the user only wanted the status code value in the series name they always have to enter a legend format to accomplish that, ex `{{status_code}}`. With this new (default) Auto mode they will not have to do that as Grafana will do that automatically. 
* To make this option efficient and not take up extra space the Legend option is by default a select with Auto, Verbose (old behavor), and Custom. If you select Custom the Select will transform into an input. 

Only new queries (with empty expr) will opt-in to this new behavior. 

* [x] Frontend tests
* [x] Backend tests 

For future consideration 
* The automatic naming could be improved (as it currently does not identify if all label values are the same in a set, and hence can be omitted) 
* The automatic naming could be moved/copied to the backend as it would be more efficient to perform there, and alerting would take advantage of it 

https://user-images.githubusercontent.com/10999/153888893-8a02c159-96d0-4b55-982a-efdb3a0eb7b1.mp4

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/querybuilder/components/PromQueryCodeEditor.tsx [hunk 1]
```diff
@@ -2,18 +2,37 @@ import React from 'react';
...
+        data-testid={testIds.editor}
+      />
+    </div>
   );
 }
+
+const getStyles = (theme: GrafanaTheme2) => {
+  return {
+    // This wrapper styling can be removed after the old PromQueryEditor is removed.
+    // This is removing margin botton on the old legacy inline form styles
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/querybuilder/components/PromQueryCodeEditor.tsx:31]:
```suggestion
    // This is removing margin bottom on the old legacy inline form styles
```

**Review by ivanahuckova — COMMENTED** [review]
I have found 1 issue. When you go from `Auto` -> `Custom`, suggested `legendFormat` is `__auto` instead of `{{label_name}}`. When I switch from `Verbose` -> `Custom` it correctly works. 

https://user-images.githubusercontent.com/30407135/154237006-f297f348-c918-405b-af0a-a112a8c976b2.mov

**Comment by torkelo** [comment]
@ivanahuckova good catch!  was due to internal useState changing faster then external (onChange), and realised the internal state was unnecessary (could be derived from query.legendFormat). Now works as it should :)

**Comment by aocenas** [comment]
Looks good, only thing I noticed and not sure if it's an issue or not is that for some queries auto is sort of more verbose than verbose:
Screenshot from 2022-02-16 11-19-31

Screenshot from 2022-02-16 11-19-38

The auto has a label `__name__` instead of just prefixing the metric name to the labels.

**Review by aocenas — APPROVED** [review]

**Comment by torkelo** [comment]
@aocenas 

Yea I noticed that as well, it's because the backend implementation is not moving the __name__ label from the labels and using it for the frame name like the frontend Prometheus result processing did before. Thinking of moving the auto naming to the Prometheus backend and improving it a bit in a future PR

**Review by ivanahuckova — APPROVED** [review]
I committed the spell check error (I hope it is okay). LGTM!

**Pull request merged**